data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconBlockHandler.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.beaconrestapi.beaconhandlers;
+
+import static tech.pegasys.artemis.provider.JsonProvider.objectToJSON;
+
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.beaconrestapi.handlerinterfaces.BeaconRestApiHandler;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+
+public class BeaconBlockHandler implements BeaconRestApiHandler {
+
+  private ChainStorageClient client;
```

Comment:
This should be marked `final`.  Class fields are the one place in Java that `final` actually make a different beyond the compiler - final fields assigned in the constructor are guaranteed to be safe to access from other threads by the time the constructor exits.  Non-final fields may have the construction rearranged so the field winds up appearing as `null` or possibly even partially constructed.
```suggestion
  private final ChainStorageClient client;
```

Reply:
done

Reply:
"Non-final fields may have the construction rearranged so the field winds up appearing as null or possibly even partially constructed." How can the construction be rearranged?

Reply:
CPUs rarely execute things in order unless there's something that tells them they absolutely have to.  From a single thread's point of view it will always appear that things happened in order (though they didn't necessarily) because the Java Memory Model guarantees it. Across threads though there's no guarantee unless there's some form of synchronisation. Instructions which appear in order to the thread they happen on may appear out of order to other threads either because the CPU really did them out of order or because the CPU memory caches had stale data.

Normally you fix that by using `synchronized` or `volatile` to add memory fences which act like checkpoints requiring a consistent view of memory. There's a special case in the Java Memory Model that guarantees an object appears fully constructed to all threads so long as all the object's fields are final and the constructor itself doesn't expose a reference to the object to another thread.

Reply:
Sorry, not all the object's fields have to be final, but the guarantees only apply to the fields that are final. http://www.cs.umd.edu/users/pugh/java/memoryModel/jsr-133-faq.html has a FAQ - the "How do final fields work under the new JMM?" question is the relevant one.

Reply:
Thank you.